# Analiza i Bazy Danych
## Karol Musiał, grupa 3a
### Laboratorium 3
## "Łączenie z bazą danych i instrukcja select"

In [2]:
import psycopg2 as pg
import pandas as pd

connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

## 1. <b>Ile kategorii filmów mamy w wypożyczalni?<b>

In [17]:
df = pd.read_sql('SELECT * FROM category',con=connection)
num_of_films = len(df.index)
print("W wypożyczalni jest {} filmów.".format(num_of_films))

W wypożyczalni jest 16 filmów


### <b>Korzystając z podtawowej wersji funkcji 'SELECT' zmierzono liczbę wierszy w tablicy 'category'.<b>


## 2. <b>Wyświetl listę kategorii w kolejności alfabetycznej.<b>

In [4]:
df = pd.read_sql('SELECT name FROM category ORDER BY name ASC',con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

### <b>Z tablicy 'category' wyświetlono kolumnę odpowiadającą nazwom kategorii oraz posortowano ją w kolejności alfabetycznej przy pomocy 'ORDER BY ... ASC'.<b>

## 3. <b>Znajdź najstarszy i najmłodszy film do wypożyczenia.<b>

In [5]:
df = pd.read_sql('SELECT * FROM film ORDER BY release_year ASC LIMIT 1',con=connection)
df

film_id            title  \
0      133  Chamber Italian   

                                         description  release_year  \
0  A Fateful Reflection of a Moose And a Husband ...          2006   

   language_id  rental_duration  rental_rate  length  replacement_cost rating  \
0            1                7         4.99     117             14.99  NC-17   

              last_update special_features  \
0 2013-05-26 14:50:58.951       [Trailers]   

                                            fulltext  
0  'chamber':1 'fate':4 'husband':11 'italian':2 ...

In [6]:
df = pd.read_sql('SELECT * FROM film ORDER BY release_year DESC LIMIT 1',con=connection)
df

film_id            title  \
0      133  Chamber Italian   

                                         description  release_year  \
0  A Fateful Reflection of a Moose And a Husband ...          2006   

   language_id  rental_duration  rental_rate  length  replacement_cost rating  \
0            1                7         4.99     117             14.99  NC-17   

              last_update special_features  \
0 2013-05-26 14:50:58.951       [Trailers]   

                                            fulltext  
0  'chamber':1 'fate':4 'husband':11 'italian':2 ...

### <b>Z tablicy 'film' wypisano posortowane po roku produkcji, pierwszy i ostatni rekord. W tej tablicy wszystkie filmy powstały w tym samym roku (2006) bez doprezycowania miesiąca i dnia produkcji, dlatego najstarszy i najmłodszy film są tym samym.<b>

## 4. <b>Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?<b>

In [20]:
df = pd.read_sql("SELECT * FROM rental WHERE rental_date BETWEEN '2005-07-01' AND '2005-08-01'",con=connection)
print("W tym okresie odbyło się {} wypożyczeń.".format(len(df.index)))

W tym okresie odbyło się 6709 wypożyczeń.


## 5. <b>Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?<b>

In [21]:
df = pd.read_sql("SELECT * FROM rental WHERE rental_date BETWEEN '2010-01-01' AND '2011-02-01'",con=connection)
print("W tym okresie odbyło się {} wypożyczeń.".format(len(df.index)))

W tym okresie odbyło się 0 wypożyczeń.


### <b>W powyższych zadaniach zastosowano składnie 'BETWEEN' i wypisano liczby rekordów w pozyskanych tablicach. W pierwszym wypadku wypożyczeń bylo 6709, natomiast w drugim nie było ich wcale.<b>

## 6. <b>Znajdź największą płatność wypożyczenia.<b>

In [9]:
df = pd.read_sql("SELECT * FROM payment ORDER BY amount DESC LIMIT 1",con=connection)
df

payment_id  customer_id  staff_id  rental_id  amount  \
0       20403          362         1      14759   11.99   

                payment_date  
0 2007-03-21 21:57:24.996577

### <b>Największą płatnośc wypożyczenia uzyskano poprzez sortowanie i ograniczenie rekordów w tablicy 'payment'. Wynosi ona '11.29'($?).<b>

## 7. <b>Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.<b>

In [22]:
# ID Polski Nigerii i Bangladeszu
df = pd.read_sql("SELECT * FROM country WHERE country in ('Poland', 'Nigeria', 'Bangladesh')",con=connection)
country_ids = tuple(df['country_id'].tolist())

# ID wszystkich miast z powyższych państw
df = pd.read_sql("SELECT * FROM city WHERE country_id in {}".format(country_ids),con=connection)
city_ids = tuple(df['city_id'].tolist())

# ID wszystkich adresów z tych miast
df = pd.read_sql("SELECT * FROM address WHERE city_id in {}".format(city_ids),con=connection)
address_ids = tuple(df['address_id'].tolist())

# Wszyscy klienci o danych adress_id
df = pd.read_sql("SELECT * FROM customer WHERE address_id in {}".format(address_ids), con=connection)
df

customer_id  store_id first_name  last_name  \
0            18         2      Carol     Garcia   
1            21         1   Michelle      Clark   
2            80         1    Marilyn       Ross   
3           103         1     Gladys   Hamilton   
4           128         1   Marjorie     Tucker   
5           156         1     Bertha   Ferguson   
6           198         2      Elsie     Kelley   
7           232         2  Constance       Reid   
8           250         2         Jo     Fowler   
9           270         1       Leah     Curtis   
10          277         2       Olga    Jimenez   
11          280         2     Tracey    Barrett   
12          284         1      Sonia    Gregory   
13          286         1      Velma      Lucas   
14          318         1      Brian      Wyman   
15          329         2      Frank   Waggoner   
16          332         1    Stephen     Qualls   
17          380         1    Russell    Brinson   
18          409         2     Rodney    Moeller   
19          501         1      Ruben      Geary   
20          558         1     Jimmie  Eggleston   
21          562         1    Wallace      Slone   
22          571         2    Johnnie   Chisholm   
23          572         1     Sidney   Burleson   

                                  email  address_id  activebool create_date  \
0       carol.garcia@sakilacustomer.org          22        True  2006-02-14   
1     michelle.clark@sakilacustomer.org          25        True  2006-02-14   
2       marilyn.ross@sakilacustomer.org          84        True  2006-02-14   
3    gladys.hamilton@sakilacustomer.org         107        True  2006-02-14   
4    marjorie.tucker@sakilacustomer.org         132        True  2006-02-14   
5    bertha.ferguson@sakilacustomer.org         160        True  2006-02-14   
6       elsie.kelley@sakilacustomer.org         202        True  2006-02-14   
7     constance.reid@sakilacustomer.org         236        True  2006-02-14   
8          jo.fowler@sakilacustomer.org         254        True  2006-02-14   
9        leah.curtis@sakilacustomer.org         275        True  2006-02-14   
10      olga.jimenez@sakilacustomer.org         282        True  2006-02-14   
11    tracey.barrett@sakilacustomer.org         285        True  2006-02-14   
12     sonia.gregory@sakilacustomer.org         289        True  2006-02-14   
13       velma.lucas@sakilacustomer.org         291        True  2006-02-14   
14       brian.wyman@sakilacustomer.org         323        True  2006-02-14   
15    frank.waggoner@sakilacustomer.org         334        True  2006-02-14   
16    stephen.qualls@sakilacustomer.org         337        True  2006-02-14   
17   russell.brinson@sakilacustomer.org         385        True  2006-02-14   
18    rodney.moeller@sakilacustomer.org         414        True  2006-02-14   
19       ruben.geary@sakilacustomer.org         506        True  2006-02-14   
20  jimmie.eggleston@sakilacustomer.org         564        True  2006-02-14   
21     wallace.slone@sakilacustomer.org         568        True  2006-02-14   
22  johnnie.chisholm@sakilacustomer.org         577        True  2006-02-14   
23   sidney.burleson@sakilacustomer.org         578        True  2006-02-14   

               last_update  active  
0  2013-05-26 14:49:45.738       1  
1  2013-05-26 14:49:45.738       1  
2  2013-05-26 14:49:45.738       1  
3  2013-05-26 14:49:45.738       1  
4  2013-05-26 14:49:45.738       1  
5  2013-05-26 14:49:45.738       1  
6  2013-05-26 14:49:45.738       1  
7  2013-05-26 14:49:45.738       1  
8  2013-05-26 14:49:45.738       1  
9  2013-05-26 14:49:45.738       1  
10 2013-05-26 14:49:45.738       1  
11 2013-05-26 14:49:45.738       1  
12 2013-05-26 14:49:45.738       1  
13 2013-05-26 14:49:45.738       1  
14 2013-05-26 14:49:45.738       1  
15 2013-05-26 14:49:45.738       1  
16 2013-05-26 14:49:45.738       1  
17 2013-05-26 14:49:45.738       1  
18 2013-05-26 14:49:45.738       1  
19 2013-05-2

### <b>Zadanie rozpoczęto od znalezienia 'country_id' Polski Nigerii i Bangladeszu, zapisano je do listy. Następnie znaleziono 'city_id' wszystkich miast w tych krajach, to samo zrobiono z 'adress_id' adresów. Ostatnim etapem było wypisanie tablicy wszystkich klientów których 'adress_id' znajduje się w liście adresów.<b>

## 8. <b>Gdzie mieszkają członkowie personelu?<b>

In [23]:
# Lista adress_id wszystkich pracowników
df = pd.read_sql("SELECT * FROM staff", con=connection)
address_ids = tuple(df['address_id'].tolist())

# Wypisanie danych adresów
df = pd.read_sql("SELECT * FROM address WHERE address_id IN {}".format(address_ids), con=connection)
df

address_id               address address2 district  city_id postal_code  \
0           3     23 Workhaven Lane     None  Alberta      300               
1           4  1411 Lillydale Drive     None      QLD      576               

         phone         last_update  
0  14033335568 2006-02-15 09:45:30  
1   6172235589 2006-02-15 09:45:30

### <b>W tablicy 'staff' znaleziono jedynie 2 pracowników, więc zapisano tylko 2 adresy. Następnie wypisano je korzystając z tablicy 'adress'.<b>

## 9. <b>Ilu pracowników mieszka w Argentynie lub Hiszpanii?<b>

In [24]:
# ID Hiszpanii i Argentyny
df = pd.read_sql("SELECT * FROM country WHERE country in ('Spain', 'Argentina')",con=connection)
country_ids = tuple(df['country_id'].tolist())

# ID miast
df = pd.read_sql("SELECT * FROM city WHERE country_id in {}".format(country_ids),con=connection)
city_ids = tuple(df['city_id'].tolist())

# ID adresów
df = pd.read_sql("SELECT * FROM address WHERE city_id in {}".format(city_ids),con=connection)
address_ids = tuple(df['address_id'].tolist())

# Tablica pracowników o danych adress_id
df = pd.read_sql("SELECT * FROM staff WHERE address_id in {}".format(address_ids), con=connection)
df

Empty DataFrame
Columns: [staff_id, first_name, last_name, address_id, email, store_id, active, username, password, last_update, picture]
Index: []

### <b>Tok postępowania był analogiczny jak w Zadaniu 7, jedyną różnicą było to, że wypisano wszystkich pracowników o danym 'adress_id' (a nie klientów). W poprzednim podpunkcie okazało się, że tablica liczy tylko 2 pracowników, nie jest więc niespodzianką że lista pracowników mieszkających w Argentynie lub Hipszanii jest pusta.<b>

## 10. <b>Jakie kategorie filmów zostały wypożyczone przez klientów?<b>

In [26]:
# Wybrano wszystkie wypożyczone kiedykolwiek inventory_id
df = pd.read_sql("SELECT DISTINCT inventory_id FROM rental ",con=connection)
inventory_ids = tuple(df['inventory_id'].tolist())

# film_id odpowiadające znalezionym inventory_id
df = pd.read_sql("SELECT * FROM inventory WHERE inventory_id in {}".format(inventory_ids),con=connection)
film_ids = tuple(df['film_id'].tolist())

# category_id odpowiadające znalezionym film_id
df = pd.read_sql("SELECT * FROM film_category WHERE film_id in {}".format(film_ids),con=connection)
category_ids = tuple(df['category_id'].tolist())

# Wszystkie kategoriie o danych category_id
df = pd.read_sql("SELECT * FROM category WHERE category_id in {}".format(category_ids),con=connection)
df

category_id         name         last_update
0             1       Action 2006-02-15 09:46:27
1             2    Animation 2006-02-15 09:46:27
2             3     Children 2006-02-15 09:46:27
3             4     Classics 2006-02-15 09:46:27
4             5       Comedy 2006-02-15 09:46:27
5             6  Documentary 2006-02-15 09:46:27
6             7        Drama 2006-02-15 09:46:27
7             8       Family 2006-02-15 09:46:27
8             9      Foreign 2006-02-15 09:46:27
9            10        Games 2006-02-15 09:46:27
10           11       Horror 2006-02-15 09:46:27
11           12        Music 2006-02-15 09:46:27
12           13          New 2006-02-15 09:46:27
13           14       Sci-Fi 2006-02-15 09:46:27
14           15       Sports 2006-02-15 09:46:27
15           16       Travel 2006-02-15 09:46:27

### <b>By odnaleść wszystkie wypożyczone kategorie filmów, rozpoczęto od znalezienia wszystkich 'inventory_id', następnie przypisano odpowiadające im 'film_id', tym z kolei 'category_id'. W ten sposób możliwe było wypisanie wszystkich kategorii które znajdowały się na liście. Okazało się (jak można się było spodziewać), że wszystkie kategorie były wypożyczone.<b>

## 11. <b>Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.<b>

In [28]:
# country_id USA
df = pd.read_sql("SELECT * FROM country WHERE country = 'United States'",con=connection)
country_id = tuple(df['country_id'].tolist())

# city_id wszystkich miast z USA
df = pd.read_sql("SELECT * FROM city WHERE country_id = {}".format(country_id[0]),con=connection)
city_ids = tuple(df['city_id'].tolist())

# adress_id wszystkich adresów z USA
df = pd.read_sql("SELECT * FROM address WHERE city_id IN {}".format(city_ids),con=connection)
address_ids = tuple(df['city_id'].tolist())

# customer_id wszystkich klientów z USA
df = pd.read_sql("SELECT * FROM customer WHERE address_id IN {}".format(address_ids),con=connection)
customer_ids = tuple(df['customer_id'].tolist())

# inventory_id wszystkich wypożyczeń klientów z USA
df = pd.read_sql("SELECT * FROM rental WHERE customer_id IN {}".format(customer_ids),con=connection)
invetory_ids = tuple(df['inventory_id'].tolist())

# film_id wszystkich inventory z poprzedniej listy
df = pd.read_sql("SELECT * FROM inventory WHERE inventory_id in {}".format(inventory_ids),con=connection)
film_ids = tuple(df['film_id'].tolist())

# category_id wszystkich kategorii o danych filmach
df = pd.read_sql("SELECT * FROM film_category WHERE film_id in {}".format(film_ids),con=connection)
category_ids = tuple(df['category_id'].tolist())

# Tablica kategorii z listy
df = pd.read_sql("SELECT * FROM category WHERE category_id in {}".format(category_ids),con=connection)
df

category_id         name         last_update
0             1       Action 2006-02-15 09:46:27
1             2    Animation 2006-02-15 09:46:27
2             3     Children 2006-02-15 09:46:27
3             4     Classics 2006-02-15 09:46:27
4             5       Comedy 2006-02-15 09:46:27
5             6  Documentary 2006-02-15 09:46:27
6             7        Drama 2006-02-15 09:46:27
7             8       Family 2006-02-15 09:46:27
8             9      Foreign 2006-02-15 09:46:27
9            10        Games 2006-02-15 09:46:27
10           11       Horror 2006-02-15 09:46:27
11           12        Music 2006-02-15 09:46:27
12           13          New 2006-02-15 09:46:27
13           14       Sci-Fi 2006-02-15 09:46:27
14           15       Sports 2006-02-15 09:46:27
15           16       Travel 2006-02-15 09:46:27

### <b>Wykonanie polecenia podzielono na 8 etapów. Znalezienie 'country_id' USA, znalezienie 'city_id' wszystkich miast z USA, znalezienie 'adress_id' wszystkich adresów z USA oraz znalezienie 'customer_id' wszystkich klientów o tych adresach. W ten sposób otrzymano listę wszystkich klientów mieszkających w USA. W kolejnych etapach zebrano 'inventory_id' wszystkich wypożyczeń tych klientów, odpowiadające im 'film_id', 'category_id' i ostatecznie wypisano wszystkie kategorie.<b>

## 12. <b>Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley<b>

In [29]:
# actor_id podanych aktorów
df = pd.read_sql("SELECT * FROM actor WHERE first_name ='Olympia' AND last_name ='Pfeiffer' OR first_name ='Julia' AND last_name ='Zellweger' OR first_name ='Ellen' AND last_name ='Presley' ",con=connection)
actor_ids = tuple(df['actor_id'].tolist())

# film_id filmów które zawierają danych aktorów
df = pd.read_sql("SELECT * FROM film_actor WHERE actor_id in {}".format(actor_ids),con=connection)
film_ids = tuple(df['film_id'].tolist())

# Wypisanie wszystkich filmów
df = pd.read_sql("SELECT title FROM film WHERE film_id in {}".format(film_ids),con=connection)
df

title
0            Badman Dawn
1        Bilko Anonymous
2   Breakfast Goldfinger
3      Caribbean Liberty
4       Casper Dragonfly
..                   ...
64             Turn Star
65            Wait Cider
66          Women Dorado
67             Won Dares
68         Wyoming Storm

[69 rows x 1 columns]

### <b>Korzystając z zagnieżdżonej konstrukcji 'WHERE' oraz operatorów logicznych, wydobyto 'actor_id' podanych w poleceniu aktorów, przypisano je do listy filmów i otrzymano listę filmów w których oni grali.<b>